components/compositing/compositor.rs

```diff
@@ -1340,6 +1354,10 @@ impl<Window: WindowMethods> IOCompositor<Window> {
             self.zoom_action = false;
         }
 
+        if self.window.get_animation_state() == windowing::AnimationState::Animating {
+            self.composite_if_necessary(CompositingReason::Animation);
+        }
+
```

Comment:
Maybe skip this change, and call hearbeat in `perform_updates`? `perform_updates` is the "heartbeart" of the compositor.

Reply:
The reason for doing this is that the main GL window is the one with vsync enabled, so the VR heartbeat needs to be called immediately after swapping the main window buffers. Hence having `composite_specific_target` calling the hearbeats between `self.window.present()` (which swaps the main window buffers) and `self.process_animations()`. If we want to prioritize the VR heartbeats over the main page's animations (which I think we do) then the heartbeats need to be in `composite_specific_target`.

Reply:
Right, I agree about doing the work after the vsync-swap-buffer.

But I'm not sure why process_animations is something we should care about? Isn't just a tick sent to the constellation?

This is the current tasks performed by Servo:

1. WR render
1. swap buffer (blocking, vsync)
1. tick animations

We are adding:

1. VR heartbeat
1. VR swap buffer (which is part of heartbeat)

You're suggesting:

1. WR render
1. Servo swap_buffer (blocking, vsync)
1. VR heartbeat
1. VR swap
1. Servo tick animations

Is that right?

Would it be basically equivalent to:

1. WR render
1. Servo Swap_buffer (blocking, vsync)
1. Servo tick animations
1. VR heartbeat
1. VR swap

If so, we can move the call to heartbeat at the end of perform_updates.

But it would be better to move the heartbeat work before the blocking swap_buffer;

1. VR heartbeat
1. WR render
1. Servo Swap_buffer (blocking, vsync)
1. VR swap
1. Servo tick animations

But that would require extracting the swap_buffer call from VR heartbeat.

Does it make sense?

Reply:
Hmm, I see what you mean, `process_animations()` isn't doing that much (a bunch of hash table lookups and channel sends) so we can probably afford to run it before the heartbeat.

Not sure why extracting the VR swap buffer from the VR heartbeat would help, moving the heartbeat into `perform_updates` would simplify things though.